### PR TITLE
feat(tokens): First functional test for tokens

### PR DESCRIPTION
**Pull request by ian-fox**
* add steps to the api tokens test to ensure there won't be a conflict, and check that the lastUsed time is updated
* fix some bugs in the tokens API
     - token.remove wasn't being called correctly
     - no longer have to delete `hash` before returning token values since the new `token.toJson` does that in the models layer.
* upgrade `sinon` to version 2

Related to https://github.com/screwdriver-cd/screwdriver/issues/532

### CODE REVIEW

**Comment by ian-fox** [comment]
Oh crap, auth test is broken. Need to fix that.

**Comment by ian-fox** [comment]
I don't *think* this should break the authorization test, but I can't be sure since some of them don't run on the local api, and the auth test fails when I run it against screwdriver-test because the changes in the middleware order aren't there yet.

Is there a way to see if that test will work other than clicking merge and seeing if the beta job fails?

File: features/step_definitions/api-token.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
+'use strict';
+
+const Assert = require('chai').assert;
+const jwt = require('jsonwebtoken');
+const request = require('../support/request');
+
+module.exports = function server() {
+    // eslint-disable-next-line new-cap
```
Comment (FenrirUnbound) [inline features/step_definitions/api-token.js]: Do you still need this?

Comment (ian-fox) [inline features/step_definitions/api-token.js]: Oh, huh, I could have sworn ESlint was complaining about it earlier. Apparently not!

**Comment by ian-fox** [comment]
Doing slightly more fiddling with `GET /auth/token` upon realizing that it doesn't work with JWT's any more with this change, the rest of the code should be the same though.

**Comment by ian-fox** [comment]
Talked to Petey, we're going to use the URL-safe base64 method, which means auth doesn't have to be touched after all. Flagging as WIP until I make that change (should be within the half hour)

File: plugins/auth/token.js [hunk 2]
```diff
@@ -16,7 +16,7 @@ module.exports = () => ({
         notes: 'Generate a JWT for use throughout Screwdriver',
         tags: ['api', 'auth', 'token'],
         auth: {
-            strategies: ['token', 'session', 'auth_token'],
+            strategies: ['session', 'auth_token'],
```
Comment (ian-fox) [inline plugins/auth/token.js]: Talked to @petey about this one, if `'token'` is listed among the strategies for this route then a JWT can be used to "refresh" itself by calling `GET /auth/token`.

Comment (ian-fox) [inline plugins/auth/token.js]: Potentially should be in its own PR? Not really really related to the rest of this one.

File: features/step_definitions/api-token.js [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+            }
+        }).then(() => this.getJwt(this.userAccessToken).then((response) => {
+            Assert.strictEqual(response.statusCode, 200);
+
+            this.newJwt = response.body.token;
+
+            return null;
+        })));
+
+    this.Then(/^a valid JWT is received that represents "calvin"$/, () => {
```
Comment (ian-fox) [inline features/step_definitions/api-token.js]: Not sure about the use of `"calvin"` here. If it's left as a regular expression cucumber complains about not having the right number of arguments in the callback. 

Ways I've seen this dealt with are using `/* eslint-disable no-unused-vars */` in `sd-step.js`, and a `switch` statement on the user with only one case for `'calvin'` in

Comment (ian-fox) [inline features/step_definitions/api-token.js]: Oh, `authorization` also has one where the regexp has `"main"` in it, I guess this is probably fine then.

File: features/api-token.feature [hunk 4]
```diff
@@ -13,14 +13,16 @@ Feature: User API Token
...
 
+
     Background:
-        Given the user "calvin" exists
+        Given "calvin" is logged in
 
-    @ignore
     Scenario: Generate New API Token
-        When a new API token is generated
+        And does not own a token named "tiger"
```
Comment (petey) [inline features/api-token.feature]: The `Background` directive is new to me, but it feels really weird in any case that Scenarios aren't using a `Given-When-Then` style.

I'd change this to `Given "calvin" does not own a token named "tiger"`

File: features/step_definitions/api-token.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
+'use strict';
+
+const Assert = require('chai').assert;
+const jwt = require('jsonwebtoken');
+const request = require('../support/request');
+
+module.exports = function server() {
+    this.Given(/^does not own a token named "([^"]*)"$/, tokenName =>
+        // Ensure there won't be a conflict: delete the token if it's already there
```
Comment (petey) [inline features/step_definitions/api-token.js]: I suggest refactoring this out somewhere and _also_ do this in an `after` hook for these tests. The expectation should be that the system is in the same "expected" state before and after tests are run.

File: plugins/auth/token.js (hunk 2 above)
Comment (petey) [inline plugins/auth/token.js]: I'm still pretty sure we shouldn't be able to refresh a JWT by passing the same JWT, unless this goes through the workflow used to generate the original JWT, e.g. check the auth provider [github] to ensure access hasn't been revoked.

File: test/plugins/tokens.test.js [hunk 6]
```diff
@@ -15,7 +18,13 @@ const getTokenMock = (token) => {
 
     mock.update = sinon.stub();
     mock.refresh = sinon.stub();
-    mock.toJson = sinon.stub().returns(Object.assign({}, token));
+    mock.toJson = sinon.stub().returns((() => {
```
Comment (petey) [inline test/plugins/tokens.test.js]: Maybe this will work?
```
mock.toJson = sinon.stub().callsFake(() => {
...
});
```

File: plugins/auth/token.js (hunk 2 above)
Comment (ian-fox) [inline plugins/auth/token.js]: I would have thought it only went through github if using the `oath` strategy.

My naive attempt of just making a `GET` request with an Authorization header borrowed from another request didn't work though, so I may be wrong.

**Comment by ian-fox** [comment]
Update June 8:
* Added cleanup in `after` hook
* Upgraded `sinon`
* General refactoring to have code be cleaner.

**Comment by ian-fox** [comment]
Oh, one moment, broke something else

File: features/step_definitions/api-token.js [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+                bearer: this.jwt
+            }
+        }).then(deleteResponse => Assert.strictEqual(deleteResponse.statusCode, 204));
+    });
+}
+
+module.exports = function server() {
+    this.After(() => {
+        if (this.testToken) {
+            return cleanup.apply(this, [this.testToken.name]);
```
Comment (petey) [inline features/step_definitions/api-token.js]: This would be cleaner with 
```
return cleanup.call(this, this.testToken.name);
```

File: features/step_definitions/api-token.js [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+        }
+
+        return null;
+    });
+
+    this.Given(/^"calvin" does not own a token named "([^"]*)"$/, tokenName =>
+        // Ensure there won't be a conflict: delete the token if it's already there
+        cleanup.apply(this, [tokenName]));
```
Comment (petey) [inline features/step_definitions/api-token.js]: Also use `call()` here.

File: features/step_definitions/api-token.js [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+            method: 'DELETE',
+            auth: {
+                bearer: this.jwt
+            }
+        }).then(deleteResponse => Assert.strictEqual(deleteResponse.statusCode, 204));
+    });
+}
+
+module.exports = function server() {
+    this.After(() => {
```
Comment (petey) [inline features/step_definitions/api-token.js]: Hooks typically go in the support dir.

https://github.com/screwdriver-cd/screwdriver/tree/master/features/support

https://github.com/cucumber/cucumber-js/blob/1.x/docs/support_files/hooks.md

Comment (ian-fox) [inline features/step_definitions/api-token.js]: Good call, created an `after-hook` file.

File: features/step_definitions/api-token.js (hunk 7 above)
Comment (ian-fox) [inline features/step_definitions/api-token.js]: Ahh, good to know about! Actually switched to using a `config` object like the other functions in `support` though, since the function got moved to `support` to be accessible to the `after` hook.

File: features/support/after-hook.js [hunk 10]
```diff
@@ -0,0 +1,24 @@
+'use strict';
+
+const sdapi = require('./sdapi');
+
+/**
+ * After hooks
+ * @return
+ */
+function afterHooks() {
+    this.After('@apitoken', () => {
```
Comment (FenrirUnbound) [inline features/support/after-hook.js]: Sorry, I may have missed the initial conversation.

If you need to clean-up, you'll have to do it in the `Before`. The `After` is not guaranteed to execute if a scenario fails. This is why [some of the other features have](https://github.com/screwdriver-cd/screwdriver/blob/7948251dea91758236d95a282a3a008fa6b930eb/features/step_definitions/authorization.js#L10) a `Before` defined in them and associated via tag.

Comment (ian-fox) [inline features/support/after-hook.js]: The one that cares about a potential conflict also calls the cleanup in `before`, my understanding is that this is just to leave the server in a cleaner state provided everything goes well.

Comment (ian-fox) [inline features/support/after-hook.js]: Specifically [here](https://github.com/screwdriver-cd/screwdriver/pull/581/files#diff-8c0434ae40099e4cb24df13dea09ccdfR9)

Comment (ian-fox) [inline features/support/after-hook.js]: It also looks like `cucumber` it will run after hooks even if a test fails.
